### PR TITLE
Add PAG Support for Stable Diffusion Inpaint Pipeline

### LINKED ISSUE #8710: Add PAG support to SD1.5
We recently integrated PAG into diffusers! See this PR [here] (https://github.com/huggingface/diffusers/pull/7944) we added PAG to SDXL

we also want to add PAG support to SD1.5 pipelines! we will need:

- [x] StableDiffusionPAGPipeline (assigned to @shauray8, PR https://github.com/huggingface/diffusers/pull/8725)
- [ ] StableDiffusionPAGImg2ImgPipeline https://github.com/huggingface/diffusers/pull/9463
- [ ] StableDiffusionPAGInpaintPipeline
- [ ] StableDiffusionControlNetPAGInpaintPipeline (https://github.com/huggingface/diffusers/pull/8875)
- [x] StableDiffusionControlNetPAGPipeline (assigned to @tuanh123789 )
- [ ] StableDiffusionControlNetPAGImg2ImgPipeline (assigned to @Bhavay-2001 https://github.com/huggingface/diffusers/pull/8864)

1. You should put it under the [pag folder](https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/pag)
2. you can use the implementation of SDXL PAG pipelines as a reference (see this PRhttps://github.com/huggingface/diffusers/pull/7944 and you can find all the sdxl pag pipelines here https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/pag)
3. you need to add AutoPipeline so that you can use this API to create it
    ```python
       AutoPipelineForImage2Image.from_pretrained(repo_id, controlnet=controlnet, enable_pag=True ...)
   ```
4. tests and docs 

If you are interested in working on this, Let me know which pipeline(s) you want to work on:)

### PR DESCRIPTION
**Pull request by darshil0805**
# What does this PR do?

Fixes #8710 Partially

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@yiyixuxu 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Hi, I have written the code and added the relevant tests and ran them, they failed in 3 cases, attached below, I'll need some assistance to test the code using images as I am GPU poor

 
Screenshot 2024-09-07 at 2 51 55 PM

### CODE REVIEW

**Comment by darshil0805** [comment]
Hi @yiyixuxu @a-r-r-o-w, following up in case you missed this, let me know how if there is anything I can do to speed up this process

**Comment by asomoza** [comment]
Hi, can you please resolve the conflicts?

File: src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py [hunk 1]
```diff
@@ -0,0 +1,1347 @@
...
+                if not isinstance(ip_adapter_image_embeds, list):
+                    raise ValueError(
+                        f"`ip_adapter_image_embeds` has to be of type `list` but is {type(ip_adapter_image_embeds)}"
+                    )
+                elif ip_adapter_image_embeds[0].ndim not in [3, 4]:
+                    raise ValueError(
+                        f"`ip_adapter_image_embeds` has to be a list of 3D or 4D tensors but is {ip_adapter_image_embeds[0].ndim}D"
+                    )
+
+    def prepare_latents(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py:671]: is this copied from the sd inpaint? if so add a `#Copied from` statement too

File: src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py [hunk 2]
```diff
@@ -0,0 +1,1347 @@
...
+
+        if return_noise:
+            outputs += (noise,)
+
+        if return_image_latents:
+            outputs += (image_latents,)
+
+        return outputs
+    
+    def _encode_vae_image(self, image: torch.Tensor, generator: torch.Generator):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py:734]: can you add a copied from?

File: src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py [hunk 3]
```diff
@@ -0,0 +1,1347 @@
...
+            ]
+            image_latents = torch.cat(image_latents, dim=0)
+        else:
+            image_latents = retrieve_latents(self.vae.encode(image), generator=generator)
+
+        image_latents = self.vae.config.scaling_factor * image_latents
+
+        return image_latents
+
+    def prepare_mask_latents(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py:748]: can you add copied from?

File: src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py [hunk 4]
```diff
@@ -0,0 +1,1347 @@
...
+        #     strength,
+        #     callback_steps,
+        #     output_type,
+        #     negative_prompt=None,
+        #     prompt_embeds=None,
+        #     negative_prompt_embeds=None,
+        #     ip_adapter_image=None,
+        #     ip_adapter_image_embeds=None,
+        #     callback_on_step_end_tensor_inputs=None,
+        #     padding_mask_crop=None,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py:1015]:
```suggestion
```

File: tests/pipelines/pag/test_pag_sd_inpaint.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        pipe_pag = pipe_pag.to(device)
+        pipe_pag.set_progress_bar_config(disable=None)
+
+        inputs = self.get_dummy_inputs(device)
+        out_pag_enabled = pipe_pag(**inputs).images[0, -3:, -3:, -1]
+
+        assert np.abs(out.flatten() - out_pag_disabled.flatten()).max() < 1e-3
+        assert np.abs(out.flatten() - out_pag_enabled.flatten()).max() > 1e-3
+    
+    def test_pag_applied_layers(self):
```
Comment (yiyixuxu) [inline tests/pipelines/pag/test_pag_sd_inpaint.py:188]: we don't need this test here (we only need to test it on one pipeline)

**Review by yiyixuxu — COMMENTED** [review]
thanks! very nice!
I left some small feedbacks!

**Review by a-r-r-o-w — COMMENTED** [review]
Could you fix the quality tests using `make quality`? Could you also add this pipeline to the PAG docs: https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/pag.md

File: src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py (hunk 3 above)
Comment (darshil0805) [reply src/diffusers/pipelines/pag/pipeline_pag_sd_inpaint.py:754]: done

**Comment by darshil0805** [comment]
```
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image
import torch


pipe.enable_model_cpu_offload()


img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url).convert("RGB")
mask_image = load_image(mask_url).convert("RGB")

prompt = "A majestic tiger sitting on a bench"


pag_scales =  [0.0, 3.0]
guidance_scales = [7.5]

grid = []
for pag_scale in pag_scales:
    for guidance_scale in guidance_scales:
        generator = torch.Generator(device="cpu").manual_seed(1)
        images = pipe(
            prompt=prompt,
            image=init_image,
            mask_image=mask_image,
            strength=0.8,
            num_inference_steps=50,
            guidance_scale=guidance_scale,
            generator=generator,
            pag_scale=pag_scale,
        ).images
        images[0]

        grid.append(images[0])

# save the grid
from diffusers.utils import make_image_grid
make_image_grid(grid, rows=len(pag_scales), cols=len(guidance_scales)).save("test_out.png")
```

I ran this on Colab and got this output, the images look exactly the same and unappealing

test_out

**Comment by darshil0805** [comment]
```
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image
import torch

device = "cuda"

# Load the image-to-image pipeline with pre-trained weights
pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-inpainting", enable_pag=True, torch_dype = torch.float16)
pipe.to(device)
pipe.enable_model_cpu_offload()


img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url).convert("RGB")
mask_image = load_image(mask_url).convert("RGB")

prompt = "A majestic tiger sitting on a bench"


pag_scales =  [0.0, 3.0]
guidance_scales = [7.5]

grid = []
for pag_scale in pag_scales:
    for guidance_scale in guidance_scales:
        generator = torch.Generator(device="cpu").manual_seed(1)
        images = pipe(
            prompt=prompt,
            image=init_image,
            mask_image=mask_image,
            strength=0.8,
            num_inference_steps=50,
            guidance_scale=guidance_scale,
            generator=generator,
            pag_scale=pag_scale,
        ).images
        images[0]

        grid.append(images[0])

# save the grid
from diffusers.utils import make_image_grid
make_image_grid(grid, rows=len(guidance_scales), cols=len(pag_scales)).save("test_out.png")
```
I was able to get this code running on Colab, this is how the outputs look:
test_out-3

**Comment by yiyixuxu** [comment]
I think CI still not passing
can you run `make style`?

**Comment by darshil0805** [comment]
> I think CI still not passing
> can you run `make style`?

I've ran it and made the change, hopefully it passes now

File: src/diffusers/pipelines/pag/__init__.py [hunk 6]
```diff
@@ -55,7 +57,6 @@
         from .pipeline_pag_sd import StableDiffusionPAGPipeline
         from .pipeline_pag_sd_3 import StableDiffusion3PAGPipeline
         from .pipeline_pag_sd_animatediff import AnimateDiffPAGPipeline
-        from .pipeline_pag_sd_img2img import StableDiffusionPAGImg2ImgPipeline
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/__init__.py:58]: is this a merge conflict? can yo uadd this line back?

Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/__init__.py:58]: also need to add pag sd inpaint here too -currently CI fail for some import error

Comment (darshil0805) [reply src/diffusers/pipelines/pag/__init__.py:58]: done

**Comment by darshil0805** [comment]
> I'm still seeing an import error https://github.com/huggingface/diffusers/actions/runs/11266431877/job/31368980510?pr=9386#step:5:250

Not sure why there is an import error, I'm able to test the script through Colab
Here's the notebook: https://colab.research.google.com/drive/1wgqDCNDlX6Fb47BMQN7GzWKtvhpounXD?usp=sharing

**Comment by yiyixuxu** [comment]
can anyone help look into the import error here?

**Comment by ParagEkbote** [comment]
Can I please contribute to fixing the import errors?

cc: @yiyixuxu

File: tests/pipelines/pag/test_pag_sd_inpaint.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        image_slice = image[0, -3:, -3:, -1]
+
+        assert image.shape == (
+            1,
+            64,
+            64,
+            3,
+        ), f"the shape of the output image should be (1, 64, 64, 3) but got {image.shape}"
+
+        expected_slice = np.array([0.8366, 0.5513, 0.6105, 0.6213, 0.6957, 0.7400, 0.6614, 0.6102, 0.5239])
```
Comment (hlky) [inline tests/pipelines/pag/test_pag_sd_inpaint.py:243]:
```suggestion
        expected_slice = np.array([0.7190, 0.5807, 0.6007, 0.5600, 0.6350, 0.6639, 0.5680, 0.5664, 0.5230])
```
Expected slice was copied from test_pag_sdxl_inpaint

**Comment by yiyixuxu** [comment]
@hlky feel free to merge if it looks good to you!

File: tests/pipelines/pag/test_pag_sd_inpaint.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        pipeline.set_progress_bar_config(disable=None)
+
+        inputs = self.get_inputs(torch_device)
+        image = pipeline(**inputs).images
+
+        image_slice = image[0, -3:, -3:, -1].flatten()
+        assert image.shape == (1, 512, 512, 3)
+        print(image_slice.flatten())
+        expected_slice = np.array(
+            [0.41385046, 0.39608297, 0.4360491, 0.26872507, 0.32187328, 0.4242474, 0.2603805, 0.34167895, 0.46561807]
```
Comment (hlky) [inline tests/pipelines/pag/test_pag_sd_inpaint.py:297]: Slow test slices need updating. Should be good to go other than that

File: tests/pipelines/pag/test_pag_sd_inpaint.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        pipeline.enable_model_cpu_offload()
+        pipeline.set_progress_bar_config(disable=None)
+
+        inputs = self.get_inputs(torch_device, guidance_scale=0.0)
+        image = pipeline(**inputs).images
+
+        image_slice = image[0, -3:, -3:, -1].flatten()
+        assert image.shape == (1, 1024, 1024, 3)
+        expected_slice = np.array(
+            [0.41597816, 0.39302617, 0.44287828, 0.2687074, 0.28315824, 0.40582314, 0.20877528, 0.2380802, 0.39447647]
```
Comment (hlky) [inline tests/pipelines/pag/test_pag_sd_inpaint.py:314]: Here too

**Review by hlky — APPROVED** [review]

**Pull request merged**